File: src/operator/runner/gigrun-files/gigdb-helper.go
package main

import (
	"context"
	"encoding/json"
	"os"

	"github.com/redis/go-redis/v9"
)

const GIG_ENV = "GIG_ENV"

var STAGE_ID = os.Getenv("STAGE_ID")

var STEP_ID = os.Getenv("STEP_ID")

const GIG_SECRETS = "GIG_SECRETS"

var STAGE_SECRETS = STAGE_ID + "_SECRETS"

var STEP_SECRETS = STEP_ID + "_SECRETS"

func init() {
	err := os.Chdir(StepEnvGet("PWD"))
	if err != nil {
		panic(err)
	}
}

func GigDb() (*redis.Client, context.Context) {
	gigdb := redis.NewClient(&redis.Options{
		Addr:     "localhost:6379",
		Password: "", // no password set
		DB:       0,  // use default DB
	})

	var gigdbCtx = context.Background()

	return gigdb, gigdbCtx
}

func CloseGigDb(gigdb *redis.Client, err error) {
	gigdb.Close()
	if err != nil && err != redis.Nil {
		panic(err)
	}
}

func envExists(key string, field string) bool {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.HExists(gigdbCtx, key, field).Result()
	CloseGigDb(gigdb, err)
	return val
}

func envGet(key string, field string) string {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.HGet(gigdbCtx, key, field).Result()
	CloseGigDb(gigdb, err)
	return val
}

func envToJson(key string) string {
	gigdb, gigdbCtx := GigDb()
	env, err := gigdb.HGetAll(gigdbCtx, key).Result()
	CloseGigDb(gigdb, err)
	envStr, _ := json.Marshal(env)
	return string(envStr)
}

func envKeys(key string) []string {
	gigdb, gigdbCtx := GigDb()
	keys, err := gigdb.HKeys(gigdbCtx, key).Result()
	CloseGigDb(gigdb, err)
	return keys
}

func envToMap(key string) map[string]string {
	gigdb, gigdbCtx := GigDb()
	envMap, err := gigdb.HGetAll(gigdbCtx, key).Result()
	CloseGigDb(gigdb, err)
	return envMap
}

func envSet(key string, field string, value string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.HSet(gigdbCtx, key, field, value).Err()
	CloseGigDb(gigdb, err)
}

func envValues(key string) []string {
	gigdb, gigdbCtx := GigDb()
	values, err := gigdb.HVals(gigdbCtx, key).Result()
	CloseGigDb(gigdb, err)
	return values
}

func GigEnvExists(field string) bool {
	return envExists(GIG_ENV, field)
}

func GigEnvGet(field string) string {
	return envGet(GIG_ENV, field)
}

func GigEnvToJson() string {
	return envToJson(GIG_ENV)
}

func GigEnvKeys() []string {
	return envKeys(GIG_ENV)
}

func GigEnvToMap() map[string]string {
	return envToMap(GIG_ENV)
}

func GigEnvSet(field string, value string) {
	envSet(GIG_ENV, field, value)
}

func GigEnvValues() []string {
	return envValues(GIG_ENV)
}

func StageEnvExists(field string) bool {
	return envExists(STAGE_ID, field)
}

func StageEnvGet(field string) string {
	return envGet(STAGE_ID, field)
}

func StageEnvToJson() string {
	return envToJson(STAGE_ID)
}

func StageEnvKeys() []string {
	return envKeys(STAGE_ID)
}

func StageEnvToMap() map[string]string {
	return envToMap(STAGE_ID)
}

func StageEnvSet(field string, val string) {
	envSet(STAGE_ID, field, val)
}

func StageEnvValues() []string {
	return envValues(STAGE_ID)
}

func StepEnvExists(field string) bool {
	return envExists(STEP_ID, field)
}

func StepEnvGet(field string) string {
	return envGet(STEP_ID, field)
}

func StepEnvToJson() string {
	return envToJson(STEP_ID)
}

func StepEnvKeys() []string {
	return envKeys(STEP_ID)
}

func StepEnvToMap() map[string]string {
	return envToMap(STEP_ID)
}

func StepEnvSet(field string, value string) {
	envSet(STEP_ID, field, value)
}

func StepEnvValues() []string {
	return envValues(STEP_ID)
}

func GigSecretsAdd(members ...string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.SAdd(gigdbCtx, GIG_SECRETS, members).Err()
	CloseGigDb(gigdb, err)
}

func GigSecretExists(field string) bool {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.SIsMember(gigdbCtx, GIG_SECRETS, field).Result()
	CloseGigDb(gigdb, err)
	return val
}

func GigSecrets() []string {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.SMembers(gigdbCtx, GIG_SECRETS).Result()
	CloseGigDb(gigdb, err)
	return val
}

func GigSecretsRemove(field string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.SRem(gigdbCtx, GIG_SECRETS, field, field).Err()
	CloseGigDb(gigdb, err)
}

func StageSecretsAdd(members ...string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.SAdd(gigdbCtx, STAGE_SECRETS, members).Err()
	CloseGigDb(gigdb, err)
}

func StageSecretExists(field string) bool {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.SIsMember(gigdbCtx, STAGE_SECRETS, field).Result()
	CloseGigDb(gigdb, err)
	return val
}

func StageSecrets() []string {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.SMembers(gigdbCtx, STAGE_SECRETS).Result()
	CloseGigDb(gigdb, err)
	return val
}

func StageSecretsRemove(field string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.SRem(gigdbCtx, STAGE_SECRETS, field, field).Err()
	CloseGigDb(gigdb, err)
}

func StepSecretsAdd(members ...string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.SAdd(gigdbCtx, STEP_SECRETS, members).Err()
	CloseGigDb(gigdb, err)
}

func StepSecretExists(field string) bool {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.SIsMember(gigdbCtx, STEP_SECRETS, field).Result()
	CloseGigDb(gigdb, err)
	return val
}

func StepSecrets() []string {
	gigdb, gigdbCtx := GigDb()
	val, err := gigdb.SMembers(gigdbCtx, STEP_SECRETS).Result()
	CloseGigDb(gigdb, err)
	return val
}

func StepSecretsRemove(field string) {
	gigdb, gigdbCtx := GigDb()
	err := gigdb.SRem(gigdbCtx, STEP_SECRETS, field, field).Err()
	CloseGigDb(gigdb, err)
}

File: .deploy/gigs-operator/runtime-support/gigdb_helper.py
import os
import json

import redis

GIG_ENV = "GIG_ENV"

STAGE_ID = os.getenv("STAGE_ID")

STEP_ID = os.getenv("STEP_ID")

GIG_SECRETS = "GIG_SECRETS"

STAGE_SECRETS = STAGE_ID + "_SECRETS"

STEP_SECRETS = STEP_ID + "_SECRETS"

gigdb = redis.Redis(host='localhost', port=6379, decode_responses=True)

def gigEnvExists(field: str):
	gigdb.hexists(GIG_ENV, field)

def gigEnvGet(field: str):
	result = gigdb.hget(GIG_ENV, field)
	return result if result else ''

def gigEnvToJson():
	gig_env = gigEnvToDict()
	return json.dumps(gig_env)

def gigEnvKeys():
	return gigdb.hkeys(GIG_ENV)

def gigEnvToDict():
	return gigdb.hegetall(GIG_ENV)

def gigEnvSet(field: str, value: str):
	return gigdb.hset(GIG_ENV, field, value)

def gigEnvValues():
	return gigdb.hvals(GIG_ENV)

def stageEnvExists(field: str):
	return gigdb.hexists(STAGE_ID, field)

def stageEnvGet(field: str):
	result = gigdb.hget(STAGE_ID, field)
	return result if result else ''

def stageEnvToJson():
	gig_env = gigEnvToDict()
	return json.dumps(gig_env)

def stageEnvKeys():
	return gigdb.hkeys(STAGE_ID)

def stageEnvToDict():
	return gigdb.hegetall(STAGE_ID)

def stageEnvSet(field: str, value: str):
	return gigdb.hset(STAGE_ID, field, value)

def stagegEnvValues():
	return gigdb.hvals(STAGE_ID)

def stepEnvExists(field: str):
	return gigdb.hexists(STEP_ID, field)

def stepEnvGet(field: str):
	result = gigdb.hget(STEP_ID, field)
	return result if result else ''

def stepEnvToJson():
	gig_env = gigEnvToDict()
	return json.dumps(gig_env)

def stepEnvKeys():
	return gigdb.hkeys(STEP_ID)

def stepEnvToDict():
	return gigdb.hegetall(STEP_ID)

def stepEnvSet(field: str, value: str):
	return gigdb.hset(STEP_ID, field, value)

def stepEnvValues():
	return gigdb.hvals(STEP_ID)

def gigSecretsAdd(*members: str):
	gigdb.sadd(GIG_SECRETS, members)

def gigSecretExists(field: str):
	return gigdb.sismember(GIG_SECRETS, field)

def gigSecrets():
	return gigdb.smembers(GIG_SECRETS)

def gigSecretsRemove(field: str):
	return gigdb.srem(GIG_SECRETS, field)

def stageSecretsAdd(*members: str):
	gigdb.sadd(STAGE_SECRETS, members)

def stageSecretExists(field: str):
	return gigdb.sismember(STAGE_SECRETS, field)

def stageSecrets():
	return gigdb.smembers(STAGE_SECRETS)

def stageSecretsRemove(field: str):
	return gigdb.srem(STAGE_SECRETS, field)

def stepSecretsAdd(*members: str):
	gigdb.sadd(STEP_SECRETS, members)

def stepSecretExists(field: str):
	return gigdb.sismember(STEP_SECRETS, field)

def stepSecrets():
	return gigdb.smembers(STEP_SECRETS)

def stepSecretsRemove(field: str):
	return gigdb.srem(STEP_SECRETS, field)
